[snapshot] Add support for concurrent simulators with Xcode 9

ohayon opened the pull request:
This PR adds support for the new XCUISnapshot API in Xcdoe 9 and running more than one simulator for screenshot generation at once
- [x] only launch 1 sim per cpu -1 (unless there is only 1 cpu) available
- [x] ensure all devices are added for xcode 9
- [x] handle OS version for devices for destination flags

**All test cases we need to check**

- [x] works in xcode 8 with 1 device
- [x] works in xcode 8 with 2 devices
- [x] works in xcode 8 with 1 lang 1 device
- [x] works in xcode 8 with 2 lang 1 device
- [x] works in xcode 8 with 1 lang 2 device
- [x] works in xcode 8 with 2 lang 2 device

- [x] works in xcode 9 with 1 device
- [x] works in xcode 9 with 2 devices
- [x] works in xcode 9 with 1 lang 1 device
- [x] works in xcode 9 with 2 lang 1 device
- [x] works in xcode 9 with 1 lang 2 device
- [x] works in xcode 9 with 2 lang 2 device
- [x] add_media(device_type, media_type, paths) work with concurrent sims

note: this will also fix: https://github.com/fastlane/fastlane/issues/9579

ohayon committed:
[WIP] [snapshot] Add support for multiple simultaneous simulators

ohayon committed:
remove some old comments

fastlane-bot-helper commented:
<!--
  0 Errors
  2 Warnings: Pull Request is Work in Progre...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Pull Request is Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You modified <code>SnapshotHelper.swift</code>, make sure to update the version number at the bottom of the file to notify users about the new helper file.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

ohayon committed:
cleanup cached screenshots

ohayon committed:
guess those comments were necessary

KrauseFx commented on snapshot/lib/snapshot/collector.rb:
```diff
@@ -26,18 +27,21 @@ def self.fetch_screenshots(output, dir_name, device_type, launch_arguments_index
         FileUtils.mkdir_p(language_folder)
 
         device_name = device_type.delete(" ")
+        # TODO: ASK FELIX ABOUT THIS!
```
Hu?

KrauseFx commented on snapshot/lib/assets/SnapshotHelper.swift:
```diff
@@ -7,7 +7,7 @@
 //
 
 // -----------------------------------------------------
-// IMPORTANT: When modifying this file, make sure to 
```
With this file being modified, will we be able to support both Xcode 8 and Xcode 9 in parallel?

taquitos replied:
The version is bumped up in this file, and there's a new `SnapshotHelperXcode8.swift`. I think when you run _snapshot_ for after updating _fastlane_ it will point you to the next `SnapshotHelperXcode8.swift`... I think, we need to test it 🍵

ohayon committed:
working to collect screenshots from multiple sims

ohayon committed:
refactoring

ohayon committed:
printing results and generating html

ohayon committed:
add more devices

Commit:
Merge branch 'master' of github.com:fastlane/fastlane into xcode9-snapshot

Commit:
Added separate paths for Xcode 8 and 9
Added SnapshotHelper for Xcode 8 projects
Dangerfile updated to handle Xcode 8 and 9 SnapshotHelper
Updated tests

Commit:
NO U

KrauseFx commented on snapshot/lib/snapshot/collector.rb:
```diff
@@ -30,7 +30,7 @@ def self.fetch_screenshots(output, dir_name, device_type, launch_arguments_index
         filename = to_store[index]
 
         device_name = device_type.delete(" ")
-        # TODO: ASK FELIX ABOUT THIS!
```
😂

KrauseFx replied:
best commit message also cc @b3ll

KrauseFx commented on fastlane.gemspec:
```diff
@@ -64,6 +64,7 @@ Gem::Specification.new do |spec|
   spec.add_dependency 'bundler', '>= 1.12.0', '< 2.0.0' # Used for fastlane plugins
   spec.add_dependency 'faraday', '~> 0.9' # Used for deploygate, hockey and testfairy actions
   spec.add_dependency 'faraday_middleware', '~> 0.9' # same as faraday
+  spec.add_dependency 'os', '~> 0.9' # Used to get CPU count for Xcode 9 multiple sim support
```
Do we need this dependency? Would there be an easy way for us to get this working without it?

KrauseFx replied:
^

taquitos replied:
Before I added os I searched through the Gemfile.lock to see if we had it transitively included due to another gem, and we did 🍾 so for right now, I think it's 👌
We could get access to this info in another way, like copying the relevant 30 lines or so from that gem 🙃

ohayon commented on snapshot/lib/snapshot/reports_generator.rb:
```diff
@@ -37,9 +37,9 @@ def generate
     end
 
-    def available_devices
+    def xcode_8_and_below_device_name_mappings
```
❤️

ohayon committed:
use new collection even if its xcode 9

ohayon committed:
refactor concurrent and single launches

ohayon committed:
renames

Commit:
Break test_command_generator into separate classes, one for future support, one for xcode 8, and a base

ohayon committed:
initialize

Commit:
Merge branch 'xcode9-snapshot' of github.com:fastlane/fastlane into xcode9-snapshot

Commit:
Fix tests
add tests
link new simulator launchers to new test command generators

Commit:
Fix more tests and add in launcher_configuration
- no reliance on globals (woo)

KrauseFx commented on snapshot/lib/assets/SnapshotHelper.swift:
```diff
@@ -33,19 +33,26 @@ func snapshot(_ name: String, waitForLoadingIndicator: Bool = true) {
 
     open class func setupSnapshot(_ app: XCUIApplication) {
+        guard let cacheDir = pathPrefix else {
+            print("Problem locating the fastlane cache directory")
```
Can we print out more information on how to resolve this?

KrauseFx commented on snapshot/lib/assets/SnapshotHelper.swift:
```diff
@@ -110,7 +109,14 @@ open class Snapshot: NSObject {
+            do {
+                try screenshot.pngRepresentation.write(to: path)
+            } catch {
+                print("Problem writing screenshot: \(name) to \(path)")
```
Can we print the actual error message here to make it easier for people to resolve the problem?

KrauseFx commented on snapshot/lib/snapshot.rb:
```diff
@@ -6,12 +6,16 @@
 require 'snapshot/fixes/simulator_zoom_fix'
 require 'snapshot/fixes/hardware_keyboard_fix'
+require 'snapshot/simulator_launchers/launcher_configuration'
+require 'snapshot/simulator_launchers/simulator_launcher'
```
Sounds like 🚀 launcher, I love it

KrauseFx commented on snapshot/lib/snapshot/collector.rb:
```diff
@@ -7,8 +7,15 @@ def self.fetch_screenshots(output, dir_name, device_type, launch_arguments_index
+      language_folder = File.join(Snapshot.config[:output_directory], dir_name)
+      FileUtils.mkdir_p(language_folder)
+
+      if Snapshot.config[:simultaneous] || Helper.xcode_at_least?(9)
```
I don't understand that `||`, what if it's Xcode 8, but `simultaneous` is true? Can this ever be the case?

taquitos replied:
This still needs to be broken down a little more because it doesn't make a ton of sense

ohayon replied:
Yeah, this is really just going to be if Xcode is at least 9, the simultaneous part doesn't matter so much anymore.

KrauseFx commented on snapshot/lib/snapshot/collector.rb:
```diff
@@ -7,8 +7,15 @@ def self.fetch_screenshots(output, dir_name, device_type, launch_arguments_index
+      if Snapshot.config[:simultaneous] || Helper.xcode_at_least?(9)
+        return collect_screenshots_for_language_folder(language_folder)
+      else
+        to_store = attachments(containing)
```
Maybe add a few lines of comment on why this `if else` block exists, and how the plist format has changes with this release

KrauseFx commented on snapshot/lib/snapshot/options.rb:
```diff
@@ -181,7 +181,17 @@ def self.available_options
+                                     is_string: false,
+                                     verify_block: proc do |value|
+                                       if value && !Helper.xcode_at_least?('9.0')
+                                         UI.user_error('Executing snapshot on multiple simulators at once is only supported in Xcode 9 and higher.')
```
Let's remove the `.` at the end of `user_error!`

ok, this line answers my other question, but then the other `if` statement is more confusing even I think

KrauseFx commented on snapshot/lib/snapshot/simulator_launchers/launcher_configuration.rb:
```diff
@@ -0,0 +1,48 @@
+    attr_accessor :output_directory
+
+    def initialize(snapshot_config: nil)
+      @languages = snapshot_config[:languages]
```
omg.... what is this, Java?

Just joking, it looks nice, but we could also use some Ruby magic if we wanted to, something along the lines of iterating over all values in `snapshot_config` and then calling the setter for `self` for each of the keys. But this is fine too, it's more explicit

taquitos replied:
I was thinking about that... that's not too magical? 🤣 

Java Beans are the best [ducks behind table]

KrauseFx commented on snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb:
```diff
@@ -0,0 +1,74 @@
+module Snapshot
+  class SimulatorLauncher < SimulatorLauncherBase
+    def default_number_of_simultaneous_simulators
+      cpu_count = OS.cpu_count
```
This will need some comments, I can imagine why we do this, but it would be good to have it documented here

taquitos replied:
I agree, we haven't linked it up yet, but that's good to call out. We did some testing and having too many sims causes serious perf issues.

KrauseFx commented on snapshot/lib/snapshot/simulator_launchers/simulator_launcher_base.rb:
```diff
@@ -0,0 +1,129 @@
+      Snapshot.kill_simulator # because of https://github.com/fastlane/snapshot/issues/337
+      `xcrun simctl shutdown booted &> /dev/null`
+
+      Fixes::SimulatorZoomFix.patch
```
It might be worth evaluating if we still need those workarounds, probably not in the scope of this PR, but something to keep in mind.

taquitos replied:
Agreed, we can follow up in another patch.

KrauseFx commented on snapshot/lib/snapshot/simulator_launchers/simulator_launcher_xcode_8.rb:
```diff
@@ -0,0 +1,107 @@
+                                             prefix: prefix_hash,
+                                            loading: "Loading...",
+                                              error: proc do |output, return_code|
+                                                ErrorHandler.handle_test_error(output, return_code)
```
I know this probably like this before, but it might be worth moving this into its own function for better readability 👍

KrauseFx replied:
☝️ optional, but something to maybe consider :)

ohayon replied:
Do you mean to extract the error block? Or the full call to `FastlaneCore::CommandExecutor`?

KrauseFx commented on snapshot/lib/snapshot/test_command_generator_base.rb:
```diff
@@ -0,0 +1,94 @@
+module Snapshot
+  class TestCommandGeneratorBase
```
I love this 😍

KrauseFx commented on snapshot/lib/snapshot/test_command_generator_base.rb:
```diff
@@ -0,0 +1,94 @@
+      end
+
+      def destination(device_name)
+        not_implemented(__method__)
```
❤️

KrauseFx commented on snapshot/lib/snapshot/test_command_generator_xcode_8.rb:
```diff
@@ -0,0 +1,63 @@
+        device = find_device(device_name, os_version)
+        if device.nil?
+          UI.user_error!("No device found named '#{device_name}' for version '#{os_version}'")
+          return
```
Unreachable code detected. `user_error!` will raise an exception

KrauseFx commented on snapshot/lib/snapshot/test_command_generator_xcode_8.rb:
```diff
@@ -0,0 +1,63 @@
+
+module Snapshot
+  # Responsible for building the fully working xcodebuild command
+  class TestCommandGeneratorXcode8 < TestCommandGeneratorBase
```
Maybe some comments here and in the other classes explaining the differences between how it works with Xcode 8 and Xcode 9

KrauseFx commented on snapshot/spec/test_command_generator_xcode_8_spec.rb:
```diff
@@ -0,0 +1,257 @@
+describe Snapshot do
+  describe Snapshot::TestCommandGeneratorXcode8 do
```
Beautiful ❤️

KrauseFx reviewed (commented):
That is an awesome PR, great work @ohayon and @taquitos, I did a quick review over the changes, I love it. 

This seems such a big change, so having a slow rollout might be a good idea, which we can easily have by launching this as soon as we merged, before Xcode 9 is out. This way we can make sure Xcode 8 keeps working as expected, and then slowly more and more people will start using Xcode 9 👍 

Let's evaluate if it's worth adding another dependency, ideally we can do it without it :) 

I didn't do any actual testing, just a review of the code.

💯 🚀 ✨

taquitos commented:
@KrauseFx 
>Let's evaluate if it's worth adding another dependency
Before I added `os` I searched through the Gemfile.lock to see if we had it transitively included due to another gem, and we did 🍾 so for right now, I think it's 👌 
We could get access to this info in another way, like copying the relevant 30 lines or so from that gem 🙃

taquitos committed:
Update test_command_generator_xcode_8.rb

remove extra return after UI.user_error!

ohayon committed:
check for device os versions and break up runs by number of cores

ohayon commented on snapshot/lib/snapshot/test_command_generator.rb:
```diff
@@ -4,15 +4,15 @@ module Snapshot
         parts += actions
         parts += suffix
-        parts += pipe(device_type, language, locale)
+        # parts += pipe(devices, language, locale)
```
I commented this out for now, we'll need to figure out how to handle logging.

Commit:
Tiny typo and downcase

Commit:
remove 'os' gem in favor of having our own implementation of cpu_count
cpu_count returns physical # of cores instead of logical (which includes hyperthreading)

ohayon committed:
test with 4 devices

ohayon committed:
write logs

ohayon commented on snapshot/lib/snapshot/test_command_generator.rb:
```diff
@@ -58,11 +58,11 @@ def verify_devices_share_os(devices)
 
-      def xcodebuild_log_path(device_type: nil, language: nil, locale: nil)
+      def xcodebuild_log_path(devices: nil, language: nil, locale: nil)
         name_components = [Snapshot.project.app_name, Snapshot.config[:scheme]]
```
Currently this is working, but the second execution (if there are more simulators than cores on the machine) will overwrite the first execution's logs. Thinking maybe we should concatenate these, but because of the `:namespace_log_files` param that is checked below, that throws a bit of a wrench into that.

ohayon committed:
append to existing logs

KrauseFx commented on snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb:
```diff
@@ -75,4 +75,30 @@ def launch_simultaneously(devices, language, locale, launch_arguments)
+          `hwprefs cpu_count`.to_i
+        when /darwin10/
+          (hwprefs_available? ? `hwprefs thread_count` : `sysctl -n hw.physicalcpu_max`).to_i
+        when /linux/
```
Does it make sense to include those if snapshot only works on macOS?

KrauseFx replied:
☝️ 👍

KrauseFx replied:
We could `UI.user_error!` if it's not macOS at this point (if we don't already to that at _snapshot_ somewhere)

Commit:
Fixes for some tests, not all

ohayon committed:
fix tests

Commit:
Delete device logs if they exist, then recreate the folder up to but excluding device log folder

Commit:
remove verbose flag for xcrun

ohayon committed:
Add missing iPads to device list

ohayon committed:
Merge branch 'xcode9-snapshot' of github.com:fastlane/fastlane into xcode9-snapshot

ohayon committed:
Add mac to device list

ohayon committed:
appleTV should use the new API too

ohayon committed:
Move operating system check up above early mac return

ohayon committed:
Add fixture for simctl output from xcode 9

ohayon committed:
fix call to add_media

ohayon committed:
update snapshot helper to throw errors with descriptions

ohayon committed:
comments

ohayon committed:
fix mac crash

ohayon committed:
default to simultaneous, provide option for serialized tests

taquitos commented on snapshot/lib/snapshot/options.rb:
```diff
@@ -182,16 +182,11 @@ def self.available_options
-        FastlaneCore::ConfigItem.new(key: :simultaneous,
-                                     env_name: "SNAPSHOT_EXECUTE_TESTS_SIMULTANEOUSLY",
-                                     description: "Take snapshots on multiple devices at once",
+        FastlaneCore::ConfigItem.new(key: :serialized_executions,
```
how about `non_concurrent_simulators` I think `serialized` is a bit confusing.
Also this property is about disabling something, so I like how `non_concurrent_simulators` sounds like it is disabling a feature.

taquitos commented on snapshot/lib/snapshot/options.rb:
```diff
@@ -182,16 +182,11 @@ def self.available_options
-                                     env_name: "SNAPSHOT_EXECUTE_TESTS_SIMULTANEOUSLY",
-                                     description: "Take snapshots on multiple devices at once",
+        FastlaneCore::ConfigItem.new(key: :serialized_executions,
+                                     env_name: "SNAPSHOT_EXECUTE_SERIALIZED_TESTS",
```
so this would be `SNAPSHOT_NON_CONCURRENT_SIMULATORS`

taquitos commented on snapshot/lib/snapshot/simulator_launchers/launcher_configuration.rb:
```diff
@@ -21,6 +21,9 @@ class SimulatorLauncherConfiguration
     attr_accessor :output_directory
 
+    # xcode 9
+    attr_accessor :serialized_executions
```
and I'd want to change this name here too, maybe make it a method `non_concurrent_simulators?`

ohayon committed:
change concurrent sims naming

ohayon committed:
rubocop

KrauseFx commented on snapshot/lib/snapshot/options.rb:
```diff
@@ -181,6 +181,11 @@ def self.available_options
+                                     is_string: false),
+        FastlaneCore::ConfigItem.new(key: :concurrent_simulators,
+                                     env_name: "SNAPSHOT_EXECUTE_CONCURRENT_SIMULATORS",
+                                     description: "Take snapshots on multiple simulators concurrently",
```
Maybe add a note to the `description` that this only works on Xcode 9 and above?

KrauseFx commented on snapshot/lib/snapshot/simulator_launchers/simulator_launcher.rb:
```diff
@@ -0,0 +1,140 @@
+
+module Snapshot
+  class SimulatorLauncher < SimulatorLauncherBase
+    def default_number_of_simultaneous_simulators
```
Maybe we can add a comment here on why we do this? Would be good for future reference

KrauseFx reviewed (approved):
Awesome PR, great work @ohayon 👍  A few smaller comments, none of them blocking 😃

KrauseFx commented:
<img width="1550" alt="screen shot 2017-08-16 at 5 52 29 pm" src="https://user-images.githubusercontent.com/869950/29391488-8bed8fc0-82ac-11e7-9b45-e8df23e8b901.png">

I tried running the most recent code from this branch, with the Xcode Beta 4 (I know, not the most recent, but I'm on a bus right now). It seems like the UI Tests failed from the snapshot example project, which is fine, but I'm a little confused on how I can find out what failed, where can I find the half-done screenshots, and where do I see the log? Did it fail on all the devices, or just some? 

Not sure if that's related to the Xcode Beta I'm using, or if it's a user error on my side :)

ohayon committed:
PR feedback

ohayon committed:
extract execution method and provide better error output

ohayon committed:
print devices that failed

ohayon committed:
working on better error handling

ohayon committed:
derp

ohayon committed:
get the first data

ohayon committed:
handle number of retries option

ohayon committed:
fix error printing on test failures

KrauseFx committed:
Small code style fix

KrauseFx committed:
Update `snapshot update` command to use Xcode 8 helper file

KrauseFx committed:
Code style improvements

taquitos reviewed (approved):
Yesssss

taquitos commented on snapshot/lib/snapshot/runner.rb:
```diff
@@ -95,15 +95,14 @@ def self.path_to_helper_file_from_gem
-      ## Something like "// SnapshotHelperVersion [1.2]", but be relaxed about whitespace
-      current_version.gsub(%r{^//\w*}, '').strip
+      # Something like "// SnapshotHelperVersion [1.2]", but be relaxed about whitespace
+      return current_version.gsub(%r{^//\w*}, '').strip
```
returns are the best 🥅      ⚽️ 🐐

taquitos commented on snapshot/lib/snapshot/update.rb:
```diff
@@ -21,7 +21,7 @@ def update
         UI.message "Updating '#{path}'..."
         input_path = Snapshot::Runner.path_to_helper_file_from_gem
-        File.write(path, File.read(input_path)) #File.read("#{Snapshot::ROOT}/lib/assets/SnapshotHelper.swift"))
+        File.write(path, File.read(input_path)) # File.read("#{Snapshot::ROOT}/lib/assets/SnapshotHelper.swift"))
```
Spaces are pretty cool too

ohayon committed:
update docs

ohayon committed:
small tweak to docs

casz commented:
Poor googlebot 💜

fishcharlie commented:
I don’t see any documentation about this. How can I enable this or use it?
